torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -182,23 +182,28 @@ void RequestCallbackImpl::processRpc(
       // scriptRemoteCall is only alive within this block, use reference to
       // avoid copy
       auto& stack = scriptRemoteCall.stackRef();
-      if (scriptRemoteCall.hasOp()) {
-        scriptRemoteCall.op()->getOperation()(stack);
-      } else {
-        PythonRpcHandler::getInstance()
-            .jitCompilationUnit()
-            ->get_function(scriptRemoteCall.qualifiedName())
-            .run(stack);
+      try {
+        if (scriptRemoteCall.hasOp()) {
+          scriptRemoteCall.op()->getOperation()(stack);
+        } else {
+          PythonRpcHandler::getInstance()
+              .jitCompilationUnit()
+              ->get_function(scriptRemoteCall.qualifiedName())
+              .run(stack);
+        }
+        TORCH_INTERNAL_ASSERT(
+            stack.size() == 1,
+            "Return value of a builtin operator or a "
+            "TorchScript function should be a single IValue, got a vector of "
+            "size ",
+            stack.size());
+        ownerRRef->setValue(std::move(stack.front()));
+      } catch (const std::exception& e) {
+        // Don't throw in this call, but rather transfer the exception
+        // to the rref.
```

Comment:
I think we should throw here to ensure that the client receives an appropriate exception (is there a reason not to?). We do attach callbacks to the Future for such messages on the client side and the error should be propagated there so that the callbacks can handle the error appropriately.

Reply:
What's the impact on he reference counting? I guess no impact, and the only difference is that the RRef is holding an error. In that case, does it mean `RemoteException` should also remove a UserRRef from `pendingUsers_`?

Reply:
This change intentionally doesn't throw here, so that the client can receive the appropriate exception.

If we throw, what happens is 
 - remoteTorchscript() call in caffe2/torch/csrc/distributed/rpc/torchscript_functions.cpp on the client side is making this call client-side
 - it runs the following:
    ctx.addPendingUser(userRRefPtr->forkId(), userRRefPtr);                                                                                                                                                          
    fm->addCallback(callback::confirmPendingUser);                                                                                                                                                                   
 - Then the confirmPendingUser call receives the exception, and blows up internally
 - The exception is never translated to the rref, and the unittest times out.

If we simply keep the flows mostly as they are, and allow the OwnerRRef to handle the exception, and propagate it to the provided future, things end up propagating as expected.

Reply:
Re: ref-counting- right, I was mostly trying to keep logic and flows as close as possible to the existing to avoid issues.

Re: pendingUsers_ - I'm trying to think, which case in particular? In this case, run the confirmPendingUser codepath just as before..

Reply:
Can we update `confirmPendingUser` to not throw an exception via `handleException` and maybe just LOG the exception for now and skip executing rest of the function in case of error? The reason I feel we should throw an error here is that in general if an RPC fails, the client should be notified of this. If the client is not handling the error properly today (ex: confirmPendingUsers), we should probably fix that.

Not throwing an exception here might be fine for now, but in the future I can imagine enhancements to the RRef protocol where we might want to be notified of this error on the client side (on the appropriate Future) and maybe have some custom handling there.

To be clear, I think we should attach the error on the RRef as we do in this PR and also throw an exception.

Reply:
> In that case, does it mean RemoteException should also remove a UserRRef from pendingUsers_?

Is this correct for the RRef protocol? If we've received an exception, we can't really remove the RRef from pendingUsers_ since we don't know if the owner has successfully recorded this particular fork.

Reply:
> The reason I feel we should throw an error here is that in general if an RPC fails, the client should be notified of this. 

There are multiple cases:

1. no error, remove the UserRRef from pendingUsers_
2. got exception here, remove the UserRRef from pendingUsers_  (as the fork has already been registered), and send back an exception message
3. got exception in other places (e.g., agent), send back an exception message

So for two, if we'd like to notify the client about the error, looks like we need to allow RemoteRet hold an exception message?

Reply:
So in this particular change, with the server-side changes, we won't see an exception on the client `confirmPendingUser` side in a server-side operator throws case like this.

(Things would be different if there's e.g. a network-related rpc failure, which is a separate issue... not sure if we should do something wildly different in that case)
 
Right now, if we throw from ``confirmPendingUser``, we end up throwing inside a future completion callback, inside (IIRC) a process_group_agent thread, and ends up being effectively logged and discarded, but in a somewhat unexpected manner.

We don't necessarily need to solve client side network error on remote issue in this particular change, but we do solve the issue of propagating the error to the rref.

Reply:
Shen,
- I agree that we should clean up the pendingUsers_ in a few corner cases. Thanks for pointing that out! I'll make those updates.

- If the server puts an exception in its ownerRRef, things already get propagated to the client's userRRef, and the exception is properly thrown, as the tests show. (But need to fix pendingUsers_)

- The case where there's e.g. a network error perceived by the client side, seems more complex.
   Right now, it appears that the only way to propagate back that async-generated error ``remoteTorchscript()`` is via the already-returned userRRef.
   It seems like there's no good [existing] mechanism for the client to place this observed network error into the userRRef, unless there's something I'm missing (?) Though I'm also concerned about widening the interface...
    What happens in real life now is that we simply time out the unittest, since the message never makes it back.